src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs

```diff
@@ -551,19 +551,24 @@ private static string EntryFromPath(string entry, int offset, int length)
             if (length == 0)
                 return String.Empty;
 
-            // create a mutable copy
-            char[] chars = entry.ToCharArray(offset, length);
-
-            // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
-            // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
-            // explicitly trying to standardize to '/'
-            for(int i = 0; i < chars.Length; i++)
+            // create a copy of the string we can modify
+            string result = String.Copy(entry);
+            
+            fixed (char* pChar = result)
             {
-                if (chars[i] == Path.DirectorySeparatorChar || chars[i] == Path.AltDirectorySeparatorChar)
-                    chars[i] = PathSeparator;
+                char* pResult = pChar; // fixed pointers are readonly
+                
+                // '/' is a more broadly recognized directory separator on all platforms (eg: mac, linux)
+                // We don't use Path.DirectorySeparatorChar or AltDirectorySeparatorChar because this is
+                // explicitly trying to standardize to '/'
+                for (int i = 0; i < result.Length; i++)
+                {
+                    if (pResult[i] == Path.DirectorySeparatorChar || pResult[i] == Path.AltDirectorySeparatorChar)
+                        pResult[i] = PathSeparator;
+                }
             }
-
-            return new string(chars);
+            
+            return result;
```

Comment:
> Don't we do this all the time in String and StringBuilder for performance reasons, though? 

Yes, but that's inside String, the same type, and StringBuilder, which is effectively an extension of that type. To code outside of such types, string should be considered immutable, except in very rare circumstances. The mutability is effectively an internal implementation detail.

> but say if we had this in a performance-critical piece of code, then what are the guidelines?

The guideline is don't do it. If there's an extremely compelling case for doing it in a particular situation, complete with scenarios and perf tests and no alternatives, then it's possible an exception would be made.